File: android/src/main/java/ee/forgr/capacitor/screenorientation/CapacitorScreenOrientationPlugin.java
package ee.forgr.capacitor.screenorientation;

import android.content.Context;
import android.content.pm.ActivityInfo;
import android.content.res.Configuration;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.view.Surface;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapacitorScreenOrientation")
public class CapacitorScreenOrientationPlugin extends Plugin implements SensorEventListener {

    private final String pluginVersion = "8.1.12";
    private int currentOrientation;
    private SensorManager sensorManager;
    private Sensor accelerometer;
    private boolean isTrackingMotion = false;
    private String currentPhysicalOrientation = "portrait-primary";
    private String lastNotifiedOrientation = null;

    @Override
    public void load() {
        super.load();
        currentOrientation = getActivity().getResources().getConfiguration().orientation;

        // Initialize sensor manager
        sensorManager = (SensorManager) getActivity().getSystemService(Context.SENSOR_SERVICE);
        if (sensorManager != null) {
            accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        }
    }

    @Override
    protected void handleOnDestroy() {
        stopMotionTracking();
        super.handleOnDestroy();
    }

    @Override
    protected void handleOnConfigurationChanged(Configuration newConfig) {
        super.handleOnConfigurationChanged(newConfig);

        // Skip system orientation changes when motion tracking is active
        // to avoid duplicate events
        if (isTrackingMotion) {
            return;
        }

        if (currentOrientation != newConfig.orientation) {
            currentOrientation = newConfig.orientation;
            notifyOrientationChange();
        }
    }

    @PluginMethod
    public void orientation(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("type", getCurrentOrientationType());
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get orientation", e);
        }
    }

    @PluginMethod
    public void lock(final PluginCall call) {
        final String orientationString = call.getString("orientation");
        final Boolean bypassLock = call.getBoolean("bypassOrientationLock", false);

        if (orientationString == null) {
            call.reject("Orientation parameter is required");
            return;
        }

        try {
            final int orientation = getOrientationConstant(orientationString);
            getActivity().setRequestedOrientation(orientation);

            // Start motion tracking if requested
            if (bypassLock) {
                startMotionTracking();
            }

            call.resolve();
        } catch (final Exception e) {
            call.reject("Could not lock orientation", e);
        }
    }

    @PluginMethod
    public void unlock(final PluginCall call) {
        try {
            stopMotionTracking();
            getActivity().setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED);
            call.resolve();
        } catch (final Exception e) {
            call.reject("Could not unlock orientation", e);
        }
    }

    @PluginMethod
    public void startOrientationTracking(final PluginCall call) {
        final Boolean bypassLock = call.getBoolean("bypassOrientationLock", false);

        if (bypassLock) {
            startMotionTracking();
        }

        call.resolve();
    }

    @PluginMethod
    public void stopOrientationTracking(final PluginCall call) {
        stopMotionTracking();
        call.resolve();
    }

    @PluginMethod
    public void isOrientationLocked(final PluginCall call) {
        try {
            final String uiOrientation = getCurrentOrientationType();
            final JSObject ret = new JSObject();

            if (isTrackingMotion) {
                // Compare physical orientation with UI orientation
                final boolean locked = !currentPhysicalOrientation.equals(uiOrientation);
                ret.put("locked", locked);
                ret.put("physicalOrientation", currentPhysicalOrientation);
                ret.put("uiOrientation", uiOrientation);
            } else {
                // No motion tracking active, can't determine if locked
                ret.put("locked", false);
                ret.put("uiOrientation", uiOrientation);
            }

            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not check orientation lock status", e);
        }
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("version", this.pluginVersion);
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get plugin version", e);
        }
    }

    private void notifyOrientationChange() {
        final JSObject ret = new JSObject();
        ret.put("type", getCurrentOrientationType());
        notifyListeners("screenOrientationChange", ret);
    }

    private String getCurrentOrientationType() {
        final int rotation = getActivity().getWindowManager().getDefaultDisplay().getRotation();
        final int orientation = getActivity().getResources().getConfiguration().orientation;

        if (orientation == Configuration.ORIENTATION_PORTRAIT) {
            return rotation == Surface.ROTATION_0 ? "portrait-primary" : "portrait-secondary";
        } else {
            return rotation == Surface.ROTATION_90 ? "landscape-primary" : "landscape-secondary";
        }
    }

    private int getOrientationConstant(final String orientationString) {
        switch (orientationString) {
            case "any":
                return ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED;
            case "natural":
                return ActivityInfo.SCREEN_ORIENTATION_NOSENSOR;
            case "landscape":
                return ActivityInfo.SCREEN_ORIENTATION_SENSOR_LANDSCAPE;
            case "portrait":
                return ActivityInfo.SCREEN_ORIENTATION_SENSOR_PORTRAIT;
            case "portrait-primary":
                return ActivityInfo.SCREEN_ORIENTATION_PORTRAIT;
            case "portrait-secondary":
                return ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT;
            case "landscape-primary":
                return ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE;
            case "landscape-secondary":
                return ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE;
            default:
                return ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED;
        }
    }

    // Motion tracking methods
    private void startMotionTracking() {
        if (isTrackingMotion || accelerometer == null) {
            return;
        }

        isTrackingMotion = true;
        sensorManager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_NORMAL);
        android.util.Log.i("ScreenOrientation", "Started motion-based orientation tracking");
    }

    private void stopMotionTracking() {
        if (!isTrackingMotion) {
            return;
        }

        isTrackingMotion = false;
        if (sensorManager != null) {
            sensorManager.unregisterListener(this);
        }
        android.util.Log.i("ScreenOrientation", "Stopped motion-based orientation tracking");
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (event.sensor.getType() != Sensor.TYPE_ACCELEROMETER) {
            return;
        }

        float x = event.values[0];
        float y = event.values[1];
        float z = event.values[2];

        // Determine orientation based on accelerometer values
        String newOrientation = determinePhysicalOrientation(x, y, z);

        if (!newOrientation.equals(currentPhysicalOrientation)) {
            currentPhysicalOrientation = newOrientation;

            // Notify listeners of physical orientation change
            if (!newOrientation.equals(lastNotifiedOrientation)) {
                lastNotifiedOrientation = newOrientation;
                final JSObject ret = new JSObject();
                ret.put("type", newOrientation);
                notifyListeners("screenOrientationChange", ret);
            }
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // Not needed for orientation detection
    }

    private String determinePhysicalOrientation(float x, float y, float z) {
        final float threshold = 5.0f;

        // X axis: left/right tilt
        // Y axis: forward/backward tilt
        // Z axis: up/down (gravity when flat)
        if (Math.abs(x) > threshold && Math.abs(x) > Math.abs(y)) {
            // Landscape orientation
            return x > 0 ? "landscape-primary" : "landscape-secondary";
        } else if (Math.abs(y) > threshold) {
            // Portrait orientation
            return y > 0 ? "portrait-primary" : "portrait-secondary";
        }
        // Default to current if unclear
        return currentPhysicalOrientation;
    }
}

File: ios/Sources/CapacitorScreenOrientationPlugin/CapacitorScreenOrientationPlugin.swift
import Foundation
import Capacitor
import CoreMotion
import UIKit

/**
 * Capacitor Screen Orientation Plugin
 *
 * Provides screen orientation detection and control with support for
 * bypassing device orientation lock using Core Motion sensors.
 */
@objc(CapacitorScreenOrientationPlugin)
public class CapacitorScreenOrientationPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.1.12"
    public let identifier = "CapacitorScreenOrientationPlugin"
    public let jsName = "CapacitorScreenOrientation"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "orientation", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "lock", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unlock", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startOrientationTracking", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopOrientationTracking", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isOrientationLocked", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private var motionManager: CMMotionManager?
    private var currentDeviceOrientation: UIDeviceOrientation = .portrait
    private var isTrackingWithMotion = false
    private var lastNotifiedOrientation: String?

    override public func load() {
        // Listen for device orientation changes from system
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.orientationDidChange),
            name: UIDevice.orientationDidChangeNotification,
            object: nil
        )

        // Start monitoring device orientation
        UIDevice.current.beginGeneratingDeviceOrientationNotifications()
    }

    deinit {
        stopMotionTracking()
        UIDevice.current.endGeneratingDeviceOrientationNotifications()
        NotificationCenter.default.removeObserver(self)
    }

    @objc private func orientationDidChange() {
        // Skip system orientation changes when motion tracking is active
        // to avoid duplicate events
        guard !isTrackingWithMotion else { return }

        let orientation = UIDevice.current.orientation
        if orientation.isValidInterfaceOrientation {
            notifyOrientationChange(fromDeviceOrientation: orientation)
        }
    }

    @objc func orientation(_ call: CAPPluginCall) {
        let orientationType = getCurrentOrientationType()
        call.resolve(["type": orientationType])
    }

    @objc func lock(_ call: CAPPluginCall) {
        guard let orientationString = call.getString("orientation") else {
            call.reject("Orientation parameter is required")
            return
        }

        let bypassLock = call.getBool("bypassOrientationLock") ?? false

        DispatchQueue.main.async {
            // Lock the orientation at the app level
            if let mask = self.getOrientationMask(from: orientationString) {
                // Note: Actual implementation would need to communicate with AppDelegate
                // to set the supported interface orientations
                // For now, we'll just track the preference
                print("Locking orientation to: \(orientationString)")
            }

            // Start motion tracking if bypass is requested
            if bypassLock {
                self.startMotionTracking()
            }

            call.resolve()
        }
    }

    @objc func unlock(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            // Stop motion tracking
            self.stopMotionTracking()

            // Unlock orientation at app level
            print("Unlocking orientation")

            call.resolve()
        }
    }

    @objc func startOrientationTracking(_ call: CAPPluginCall) {
        let bypassLock = call.getBool("bypassOrientationLock") ?? false

        if bypassLock {
            startMotionTracking()
        }

        call.resolve()
    }

    @objc func stopOrientationTracking(_ call: CAPPluginCall) {
        stopMotionTracking()
        call.resolve()
    }

    @objc func isOrientationLocked(_ call: CAPPluginCall) {
        let uiOrientation = getCurrentOrientationType()

        if isTrackingWithMotion {
            // We have motion data, compare physical vs UI orientation
            let physicalOrientation = mapDeviceOrientationToString(currentDeviceOrientation)
            let locked = physicalOrientation != uiOrientation

            call.resolve([
                "locked": locked,
                "physicalOrientation": physicalOrientation,
                "uiOrientation": uiOrientation
            ])
        } else {
            // No motion tracking active, can't determine if locked
            // Return false by default, but note that we don't have physical orientation data
            call.resolve([
                "locked": false,
                "uiOrientation": uiOrientation
            ])
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }

    // MARK: - Core Motion Tracking

    private func startMotionTracking() {
        guard !isTrackingWithMotion else { return }

        motionManager = CMMotionManager()
        guard let motionManager = motionManager else { return }

        motionManager.accelerometerUpdateInterval = 0.2
        motionManager.gyroUpdateInterval = 0.2

        if motionManager.isAccelerometerAvailable {
            isTrackingWithMotion = true

            motionManager.startAccelerometerUpdates(to: .main) { [weak self] (data, _) in
                guard let self = self, let data = data else { return }

                let acceleration = data.acceleration
                let orientation = self.determineOrientation(from: acceleration)

                if orientation != self.currentDeviceOrientation {
                    self.currentDeviceOrientation = orientation
                    self.notifyOrientationChange(fromDeviceOrientation: orientation)
                }
            }

            print("Started motion-based orientation tracking")
        } else {
            print("Accelerometer not available")
        }
    }

    private func stopMotionTracking() {
        guard isTrackingWithMotion else { return }

        motionManager?.stopAccelerometerUpdates()
        motionManager = nil
        isTrackingWithMotion = false

        print("Stopped motion-based orientation tracking")
    }

    private func determineOrientation(from acceleration: CMAcceleration) -> UIDeviceOrientation {
        let threshold = 0.5

        if acceleration.x < -threshold {
            return .landscapeRight
        } else if acceleration.x > threshold {
            return .landscapeLeft
        } else if acceleration.y < -threshold {
            return .portrait
        } else if acceleration.y > threshold {
            return .portraitUpsideDown
        }

        return currentDeviceOrientation
    }

    // MARK: - Helper Methods

    private func getCurrentOrientationType() -> String {
        let interfaceOrientation = UIApplication.shared.windows.first?.windowScene?.interfaceOrientation ?? .portrait
        return mapInterfaceOrientationToString(interfaceOrientation)
    }

    private func notifyOrientationChange(fromDeviceOrientation deviceOrientation: UIDeviceOrientation) {
        let orientationType = mapDeviceOrientationToString(deviceOrientation)

        // Only notify if orientation actually changed
        if orientationType != lastNotifiedOrientation {
            lastNotifiedOrientation = orientationType
            notifyListeners("screenOrientationChange", data: ["type": orientationType])
        }
    }

    private func mapInterfaceOrientationToString(_ orientation: UIInterfaceOrientation) -> String {
        switch orientation {
        case .portrait:
            return "portrait-primary"
        case .portraitUpsideDown:
            return "portrait-secondary"
        case .landscapeLeft:
            return "landscape-primary"
        case .landscapeRight:
            return "landscape-secondary"
        default:
            return "portrait-primary"
        }
    }

    private func mapDeviceOrientationToString(_ orientation: UIDeviceOrientation) -> String {
        switch orientation {
        case .portrait:
            return "portrait-primary"
        case .portraitUpsideDown:
            return "portrait-secondary"
        case .landscapeLeft:
            return "landscape-secondary" // Note: device left = interface right
        case .landscapeRight:
            return "landscape-primary" // Note: device right = interface left
        default:
            return "portrait-primary"
        }
    }

    private func getOrientationMask(from orientationString: String) -> UIInterfaceOrientationMask? {
        switch orientationString {
        case "any":
            return .all
        case "natural", "portrait":
            return .portrait
        case "landscape":
            return .landscape
        case "portrait-primary":
            return .portrait
        case "portrait-secondary":
            return .portraitUpsideDown
        case "landscape-primary":
            return .landscapeLeft
        case "landscape-secondary":
            return .landscapeRight
        default:
            return nil
        }
    }
}

// Helper extension
extension UIDeviceOrientation {
    var isValidInterfaceOrientation: Bool {
        switch self {
        case .portrait, .portraitUpsideDown, .landscapeLeft, .landscapeRight:
            return true
        default:
            return false
        }
    }
}
